### PR TITLE
python coverage integration

### PR DESCRIPTION
**Pull request by pylipp**
This PR integrates Python code coverage for both local and CI runs.

CircleCI does not support generating code-coverage badges, so we'd have to run the code-coverage on e.g. coveralls. However I don't have the rights to add the boxtribute repo there. Instructions are [here](https://circleci.com/docs/2.0/code-coverage/#coveralls).

### CODE REVIEW

**Comment by aerinsol** [comment]
> This PR integrates Python code coverage for both local and CI runs.
> 
> CircleCI does not support generating code-coverage badges, so we'd have to run the code-coverage on e.g. coveralls. However I don't have the rights to add the boxtribute repo there. Instructions are [here](https://circleci.com/docs/2.0/code-coverage/#coveralls).

Looks like we have codecov permissions, which is the more stable service. https://www.slant.co/versus/7928/7929/~coveralls_vs_codecov Instructions to integrate it are right above the coveralls instructions in your link. Let me know if you need the integration tokens.

**Review by aerinsol — CHANGES_REQUESTED** [review]
Integrate with codecov.io (Roanna can share tokens if need be)

**Review by aerinsol — APPROVED** [review]

File: .circleci/config.yml [hunk 1]
```diff
@@ -109,15 +110,19 @@ jobs:
...
-            source env/bin/activate
-            mkdir test-results
-            pytest --junitxml=test-results/junit.xml
+            source flask/env/bin/activate
+            mkdir flask/test-results
+            pytest --cov --cov-report=html --junitxml=flask/test-results/junit.xml flask
       - store_test_results:
-          path: test-results
+          path: flask/test-results
+      - store_artifacts:
```
Comment (HaGuesto) [inline .circleci/config.yml:118]: u are storing the artifacts twice?

File: README.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 [![CircleCI](https://circleci.com/gh/boxwise/boxtribute.svg?style=svg)](https://circleci.com/gh/boxwise/boxtribute)
 <a width="105" height="35" href="https://auth0.com/?utm_source=oss&utm_medium=gp&utm_campaign=oss" target="_blank" alt="Single Sign On & Token Based Authentication - Auth0">
 <img width="105" height="35" alt="JWT Auth for open source projects" src="https://cdn.auth0.com/oss/badges/a0-badge-dark.png"></a>
+[![codecov](https://codecov.io/gh/boxwise/boxtribute/branch/master/graph/badge.svg?token=646MWM6V9H)](https://codecov.io/gh/boxwise/boxtribute)
```
Comment (HaGuesto) [inline README.md:4]: from the link I assume that this badge only works on the master branch, right?

**Review by HaGuesto — COMMENTED** [review]
## Code review
generally looks good :) thanks for directly adding stuff to the README.

more in inline comments.

## Functional Review

Locally, I can produce codecov reports. 
I can see that the same reports are added on CircleCI as artifacts.

[Codecov.io does show errors though](https://codecov.io/gh/boxwise/boxtribute/commit/0c318cb97d06e7904e5cf3500b83b60961052167/). I did take a look at the settings and it seems like the codecov token was not set yet as environment variable. I took the liberty and added it in CircleCI. I wanted to rerun it then, but was not able to upload the [results to codecov.io](https://app.circleci.com/pipelines/github/boxwise/boxtribute/233/workflows/fb953960-b723-42b1-9410-a0db27d31d06/jobs/1019)

@pylipp I'm happy to merge, but I am not sure if you finished this PR already regarding codecov.io.

**Comment by aerinsol** [comment]
[add code coverage with pytest #small #python](https://trello.com/c/cUJbJI9t/684-add-code-coverage-with-pytest-small-python)

File: .circleci/config.yml (hunk 1 above)
Comment (pylipp) [reply .circleci/config.yml:118]: What do you mean exactly? As previously, the `test-results/` directory is stored via `store_test_results` and `store_artifacts`. Now additionally `flask/htmlcov` is stored (but not required anymore when the codecov integration works).

**Comment by pylipp** [comment]
> [Codecov.io does show errors though](https://codecov.io/gh/boxwise/boxtribute/commit/0c318cb97d06e7904e5cf3500b83b60961052167/). I did take a look at the settings and it seems like the codecov token was not set yet as environment variable. I took the liberty and added it in CircleCI. I wanted to rerun it then, but was not able to upload the [results to codecov.io](https://app.circleci.com/pipelines/github/boxwise/boxtribute/233/workflows/fb953960-b723-42b1-9410-a0db27d31d06/jobs/1019)
> 
> @pylipp I'm happy to merge, but I am not sure if you finished this PR already regarding codecov.io.

The codecov token shouldn't be required in public repos, thanks for adding the token though.

The error shown in codecov is probably due to path fixing in the coverage report, I couldn't investigate further yet.

**Comment by pylipp** [comment]
@HaGuesto can you merge this? @spaudanjo fixed the integration with codecov, and the display of the badges.
Only thing we we're wondering about: why is junit-output from pytest necessary in the CircleCI job?

**Comment by spaudanjo** [comment]
> @HaGuesto can you merge this? @spaudanjo fixed the integration with codecov, and the display of the badges.
> Only thing we we're wondering about: why is junit-output from pytest necessary in the CircleCI job?

Just to add on Philipps comment: the junit output format is not necessary for the codecov badge or the current circleCI pipeline, but maybe it was intentionally there for some other reason?

**Comment by HaGuesto** [comment]
No reason I can recollect :)

Thanks @spaudanjo and @pylipp looks good and respect for the setup.cfg

**Pull request merged**